File: Data Structures & Algorithms/buy-and-sell-crypto/submission-0.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        left = 0
        currProfit = 0
        right = 1 
        while right < len(prices):
            if prices[left] > prices[right]:
                left += 1
                if left >= right:
                    right +=1
            else:
                currProfit = max(currProfit, prices[right] - prices[left])
                right += 1
        
        return currProfit
                

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        int[] result = { 0 };

        diameterOfBinaryTreeHelper(root, result);

        return result[0];
    }

    public void diameterOfBinaryTreeHelper(TreeNode root, int[] result) {
        if (root == null) {
            return;
        }
        int heightLeft = maxHeight(root.left);
        int heightRight = maxHeight(root.right);
        result[0] = Math.max(heightLeft + heightRight, result[0]);
        diameterOfBinaryTreeHelper(root.left, result);
        diameterOfBinaryTreeHelper(root.right, result);
    }

    public int maxHeight(TreeNode root) {
        if (root == null) {
            return 0;
        }

        return 1 + Math.max(maxHeight(root.left), maxHeight(root.right));
    }
}
